Documentation/rfc/v3api.md

```diff
@@ -0,0 +1,476 @@
+  // and generates one event in the event history.
+  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
+
+  // Tnx processes all the requests in one transaction.
+  // A tnx request increases the index of the store,
+  // and generates one event in the event history.
+  rpc Tnx(TnxRequest) returns (TnxResponse) {}
+
+  // Watch watches the events happening or happened in etcd. Both input and output
+  // are stream. One watch rpc can watch for multiple ranges and get a stream of
+  // events. The whole events history can be watched unless compacted.
+  rpc Watch(stream WatchRequest) returns (stream WatchResponse) {}
+
+  // Compact compacts the event history in etcd. User should compact the
+  // event history periodically, or it will grow infinitely.
+  rpc Compact(CompactRequest) returns (CompactResponse) {}
+
+  // LeaseCreate creates a lease. A lease has a TTL. The lease will expire if the
+  // server does not receive a keepAlive within TTL from the lease holder.
+  // All keys attached to the lease will be expired and deleted if the lease expires.
+  // The key expiration generates an event in event history.
+  rpc LeaseCreate(LeaseCreateRequest) returns (LeaseCreateResponse) {}
+
+  // LeaseRevoke revokes a lease. All the key attached to the lease will be expired and deleted.
+  rpc LeaseRevoke(LeaseRevokeRequest) returns (LeaseRevokeResponse) {}
+
+  // LeaseAttach attaches keys with a lease.
+  rpc LeaseAttach(LeaseAttachRequest) returns (LeaseAttachResponse) {}
+
+  // LeaseTnx likes Tnx. It does one additional thing that all the keys put into the
+  // store are attached with the given lease. This is rpc is useful when you want to
+  // put a key and attach it to a lease atomically.
+  rpc LeaseTnx(LeaseTnxRequest) returns (LeaseTnxResponse) {}
+
+  // KeepAlive keeps the lease alive.
+  rpc LeaseKeepAlive(stream LeaseKeepAliveRequest) returns (stream LeaseKeepAliveResponse) {}
+}
+
+message RequestHeader {
+  // the cluster id to check with
+  // if the cluster id is not matched, an error is returned.
+  optional uint64 cluster_id = 1;
+}
+
+message ResponseHeader {
+  // an error type message?
+  optional string error = 1;
+  optional uint64 cluster_id = 2;
+  optional uint64 member_id = 3;
+  // index of the store when the request was applied.
+  optional int64 index = 4;
+  // term of raft when the request was applied.
+  optional uint64 raft_term = 5;
+}
+
+message RangeRequest {
+  optional bytes key = 1;
+  // if the end_key is not given, it is a get.
+  // if the end_key is given, it gets the keys in range [key, end_key).
+  optional bytes end_key = 2;
+  // limit the number of keys returned.
+  optional int64 limit = 3;
+  // if not all the keys are returned, a continue id will be 
+  // returned in the range response if consistent is set.
+  // the returned continue id only works with the previous 
+  // etcdserver the client connected to. 
+  optional bool consistent = 4
```

Comment:
> On Apr 21, 2015, at 3:58 PM, Xiang Li <notifications@github.com> wrote:
> 
> In Documentation/rfc/v3api.md:
> 
> > +  // term of raft when the request was applied.
> > +  optional uint64 raft_term = 5;
> > +}
> > +
> > +message RangeRequest {
> > +  optional bytes key = 1;
> > +  // if the end_key is not given, it is a get.
> > +  // if the end_key is given, it gets the keys in range [key, end_key).
> > +  optional bytes end_key = 2;
> > +  // limit the number of keys returned.
> > +  optional int64 limit = 3;
> > +  // if not all the keys are returned, a continue id will be 
> > +  // returned in the range response if consistent is set.
> > +  // the returned continue id only works with the previous 
> > +  // etcdserver the client connected to. 
> > +  optional bool consistent = 4
> +1000. i really think this is an uncommon case. and we should separate it and not try to optimize it at all.
> 
It's definitely separate... but it seems like this is the way most ranging clients would work above 100-500 keys - fetch in batches, with the client paging in requests as the client iterates.  A non trivial number of range queries will span a set of keys for which the request could easily span multiple megabytes (10k keys with 10k data each seems like a small range query for a store that is handling 100k or 1mil keys).  I'd be hard pressed to name a store that doesn't batch results at the driver level.

I think the snapshot query + range watch is the strongest feature of etcd - the one which distinguishes it from practically every other store out there.  I'd hate to unintentionally rule that out early unless the compromises are too great.  Imagine someone using etcd at the 10k key / 10k value level.  When they move to the next tier of sizes, do they need to rearchitect their entire query strategy with a different, more complex pattern?
> —
> Reply to this email directly or view it on GitHub.
>